src/librbd/io/ImageRequestWQ.cc

```diff
@@ -320,14 +320,32 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
     c->set_event_notify(true);
   }
 
-  if (!start_in_flight_io(c)) {
+ if (!start_in_flight_io(c)) {
     return;
   }
 
+  auto tid = ++m_last_tid;
+
+  if (m_image_ctx.non_blocking_aio || writes_blocked()) {
```

Comment:
This all still seems racy to me. I think you should just unconditionally execute `ImageDispatchSpec<I>::create_XYZ`, then queue it for processing (remove the current logic which attempts to execute it in the same thread context immediately). 

After creating the request but prior to queueing it, record the request tid in the `m_queued_or_blocked_io_tids` collection. It seems like then you could avoid the need to `prepare_overlapping_io` since you could move that logic to the process path and only block IOs if overlapping IOs are actively being dispatched instead of blocking on queued IOs as well.

Reply:
hmm, if `ImageDispatchSpec<I>::create_XYZ` is executed unconditionally and down below by the time it reaches to `m_image_ctx.non_blocking_aio` and if it's false, wouldn't that lead to a memory leak?

With the conditional check now, it only ever creates a request where it can't be queued immediately (i.e. when it needs to be blocked). In other words, if this is executed:

```
if (blocked) {
  req = ImageDispatchSpec<I>::create_write_request(
              m_image_ctx, c, {{off, len}}, std::move(bl), op_flags, trace, tid);
}
```
then it returns (after `prepare_overlapping_io`) & won't reach the below lines:

```
 std::shared_lock owner_locker{m_image_ctx.owner_lock};
 if (m_image_ctx.non_blocking_aio || writes_blocked()) {
    queue(ImageDispatchSpec<I>::create_write_request(
```

Reply:
Additionally, if `ImageDispatchSpec<I>::create_XYZ` is executed w/o a condition, and if the request is a blocking IO (`non_blocking_aio=false`), then it seems `ImageRequest<I>::aio_write` can't access the bufferlist due to `std::move` semantics

Reply:
The bug right now is that you test `m_image_ctx.non_blocking_aio || writes_blocked` under two different `if` conditions so the results might change. This will either leave `req` uninitialized when it attempts to queue it or it will leak memory if it directly executes it. 

In the cause where you would normally directly invoke it, just directly process the pre-created request.

Reply:
> This will either leave req uninitialized when it attempts to queue

In the current code, `req` variable is not used in `queue`. It is executed directly. However I do see that, if the first if  condition `m_image_ctx.non_blocking_aio` is false and the second if condition is true, code won't be checking for overlapping IOs for that request.

> In the cause where you would normally directly invoke it, just directly process the pre-created request.

I will still need to execute `ImageDispatchSpec<I>::create_XYZ` under `m_image_ctx.non_blocking_aio || writes_blocked` condition. Perhaps something like below?

```
if (m_image_ctx.non_blocking_aio || writes_blocked()) {
  ImageDispatchSpec<I> *req = ImageDispatchSpec<I>::create_write_request(
              m_image_ctx, c, {{off, len}}, std::move(bl), op_flags, trace, tid);

  if (!prepare_overlapping_io(blocked, req)) {
      return;
  }
  std::shared_lock owner_locker{m_image_ctx.owner_lock};
  queue(req);
```

Reply:
Wouldn't that just break the `else` path where it's directly executed? i.e. it might execute when it should have been blocked?

Reply:
oh you meant something like below?

```
ImageDispatchSpec<I> *req = ImageDispatchSpec<I>::create_write_request(
              m_image_ctx, c, {{off, len}}, std::move(bl), op_flags, trace, tid);

// add logic for tids into `m_queued_or_blocked_io_tids` and handle in `process`

std::shared_lock owner_locker{m_image_ctx.owner_lock};
if (m_image_ctx.non_blocking_aio || writes_blocked()) {
  queue(req);
else {
  ...
  req->send(); 
  ....
}
```

Reply:
Yup -- but in a perfect world you would be able to call `process` instead of `req->send()` (or both `process` and this could call the same helper

Reply:
Done